src/Avalonia.Base/Media/Fonts/Tables/Cmap/CmapFormat4Table.cs

```diff
@@ -0,0 +1,316 @@
+        private readonly ReadOnlyMemory<byte> _idRangeOffsets;
+        private readonly ReadOnlyMemory<byte> _glyphIdArray;
+
+        private int? _count;
+
+        public CmapFormat4Table(ReadOnlyMemory<byte> table)
+        {
+            var reader = new BigEndianBinaryReader(table.Span);
+
+            ushort format = reader.ReadUInt16(); // must be 4
+
+            Debug.Assert(format == 4, "Format must be 4.");
+
+            ushort length = reader.ReadUInt16(); // length in bytes of this subtable
+
+            _table = table.Slice(0, length);
+
+            ushort language = reader.ReadUInt16(); // language code, 0 for non-language-specific
+
+            ushort segCountX2 = reader.ReadUInt16(); // 2 * segCount
+            _segCount = segCountX2 / 2;
+
+            ushort searchRange = reader.ReadUInt16(); // searchRange = 2 * (2^floor(log2(segCount)))
+            ushort entrySelector = reader.ReadUInt16(); // entrySelector = log2(searchRange/2)
+            ushort rangeShift = reader.ReadUInt16(); // rangeShift = segCountX2 - searchRange
+
+            // Spec sanity checks
+            Debug.Assert(searchRange == (ushort)(2 * (1 << (int)Math.Floor(Math.Log(_segCount, 2)))),
+                "searchRange must equal 2 * (2^floor(log2(segCount))).");
+            Debug.Assert(entrySelector == (ushort)Math.Floor(Math.Log(_segCount, 2)),
+                "entrySelector must equal log2(searchRange/2).");
+            Debug.Assert(rangeShift == (ushort)(segCountX2 - searchRange),
+                "rangeShift must equal segCountX2 - searchRange.");
+
+            // Compute offsets
+            int endCodeOffset = reader.Position;
+            int startCodeOffset = endCodeOffset + _segCount * 2 + 2; // + reservedPad
+            int idDeltaOffset = startCodeOffset + _segCount * 2; // after startCodes
+            int idRangeOffsetOffset = idDeltaOffset + _segCount * 2; // after idDeltas
+            int glyphIdArrayOffset = idRangeOffsetOffset + _segCount * 2; // after idRangeOffsets
+
+            // Ensure declared length is consistent
+            Debug.Assert(length >= glyphIdArrayOffset,
+                "Subtable length must be at least large enough to contain glyphIdArray.");
+
+            // Slice directly
+            _endCodes = _table.Slice(endCodeOffset, _segCount * 2);
+
+            _startCodes = _table.Slice(startCodeOffset, _segCount * 2);
+
+            _idDeltas = _table.Slice(idDeltaOffset, _segCount * 2);
+
+            _idRangeOffsets = _table.Slice(idRangeOffsetOffset, _segCount * 2);
+
+            int glyphCount = (length - glyphIdArrayOffset) / 2;
+
+            Debug.Assert(glyphCount >= 0, "GlyphIdArray length must not be negative.");
+
+            _glyphIdArray = _table.Slice(glyphIdArrayOffset, glyphCount * 2);
+        }
+
+        // Reads a big-endian UInt16 from the specified word index in the given memory
+        private static ushort ReadUInt16BE(ReadOnlyMemory<byte> mem, int wordIndex)
```

Comment:
Pass a `ReadOnlySpan` here for performance and make the callers pass the span instead (see comment below).